### PR TITLE
Feat/badge system UI

### PR DESCRIPTION
**Pull request by naderbahri**
[Assisted by AI]

## Description
This PR introduces a comprehensive badge and achievement system to the MapIn profile screen, allowing users to view their unlocked achievements and track progress toward locked badges.
It adds a new BadgesSection component with complete UI implementation (grid display, detail dialogs, rarity legend), state management, accessibility support, and extensive test coverage.

**Related Issue:** #313 
**Comment:** The pull request is beyond 909 lines added because of the review's requested changed. Before, it was less than 800 lines as usual.

---

### Data Models
Added badge data structures to support achievement tracking:

- **Badge.kt:** Data class with id, title, description, icon, rarity, isUnlocked, and progress (0.0-1.0) properties
- **BadgeRarity enum:** Four difficulty levels (COMMON, RARE, EPIC, LEGENDARY) with corresponding color gradients
- **SampleBadges.kt:** Utility object providing three representative badges for development and testing

### UI / UX
Created complete badge display system:

- **BadgesSection composable:**
  - Displays achievement header with unlock counter (e.g., "2/4")
  - 4-column LazyVerticalGrid with 250dp fixed height for badge display
  - Integrated BadgeRarityLegend explaining color-coded difficulty levels
  - Click handling to open detail dialogs
  - Full Material Design 3 styling with 0dp elevation cards
  - Comprehensive KDoc documentation

- **BadgeItem component:**
  - Circular icon display (56dp) with Material Icons or lock icon for locked badges
  - Animated alpha transition (300ms tween) between locked (0.4f) and unlocked (1f) states
  - Rarity-based radial gradient backgrounds for unlocked badges
  - Border styling with semi-transparent rarity colors (0.8f alpha)
  - Progress indicator bar (3dp height) for locked badges with >0 progress
  - Two-line title text with ellipsis overflow

- **BadgeDetailDialog:**
  - Full-screen AlertDialog with 80dp circular icon
  - Badge title, rarity label (titlecased), and description
  - State-dependent content display:
    - Progress section with percentage and full-width progress bar (8dp height)
    - "Unlocked" text in rarity color for completed achievements
    - "Locked" text in muted color for zero-progress badges
  - Rarity-colored close button with 24dp rounded corners

- **BadgeRarityLegend:**
  - Educational component with "Rarity Levels" header
  - Horizontal row displaying four rarity levels with 12dp circular color dots
  - Titlecased names (Common, Rare, Epic, Legendary)

- **Visual Design System:**
  - COMMON: Blue-grey gradient (0xFF78909C to 0xFF546E7A)
  - RARE: Blue gradient (0xFF42A5F5 to 0xFF1E88E5)
  - EPIC: Purple gradient (0xFFAB47BC to 0xFF8E24AA)
  - LEGENDARY: Gold/orange gradient (0xFFFFB300 to 0xFFF57C00)

### Profile Integration
Updated ProfileScreen.kt:

- Added BadgesSection to ViewProfileContent after hobbies card
- Uses SampleBadges.getSampleBadges() for demonstration data
- Positioned at bottom of scrollable profile content with 16dp top spacing
- Appears only in view mode, not in edit mode

### Accessibility
All badge states include descriptive contentDescription:

- Badge items: Badge title for icon content description
- Visual states: Lock icons and unlock states clearly indicated
- Dialog content: Clear hierarchy with headline, labels, and body text
- Test tags: Comprehensive testTag modifiers on all interactive elements

---

### Unit Tests (6) – BadgeTest.kt
- **Default values validation:** Confirms isUnlocked = false and progress = 0f when not explicitly set
- **Data class operations:** Tests copy, equals, hashCode, toString implementations
- **Rarity enum validation:** Verifies exactly 4 values with correct names (COMMON, RARE, EPIC, LEGENDARY)
- **Progress boundaries:** Tests edge cases with minimum (0f) and maximum (1f) progress values
- **Sample data integrity:** Validates SampleBadges.getSampleBadges() returns 3 badges with correct properties
- **Code quality:** Replaced deprecated Enum.values() with Enum.entries for Kotlin 1.9+ compatibility

### Instrumentation Tests (12) – BadgeSectionTest.kt

**BadgesSection Tests (5):**
- Basic rendering with "Achievements" header and section visibility
- Badge count display with mixed locked/unlocked ratios (e.g., "2/4")
- Complete badge grid rendering with all titles and test tags
- Empty list handling with "0/0" counter
- Rarity legend display with all four rarity names

**BadgeDetailDialog Tests (2):**
- Progress display for locked badges (75% example)
- Dialog dismiss functionality via close button

**Multiple Badges Tests (3):**
- All four rarity levels display correctly with proper colors
- Different progress levels (0f, 0.25f, 0.5f, 0.75f) render accurately
- Mixed locked/unlocked scenarios with correct counter calculation

**Edge Cases (2):**
- Zero progress displays "Locked" text without progress bar
- Full progress (100%) displays percentage in dialog

**Test Cleanup:**
- Removed 6 redundant tests to improve maintainability

---

## Testing
- [x] Unit tests pass
- [x] Instrumentation tests pass
- [x] Manual testing completed
- [x] Coverage maintained (≥80%)

**Summary:**
19 tests added (6 unit + 12 instrumentation), covering all state branches, user interactions, and data model behavior.
State-to-UI flow fully validated using sample data and MockK for ImageVector mocking.

### Coverage:
80%+ coverage across:
- Badge data class and BadgeRarity enum
- SampleBadges utility
- BadgesSection component (grid, header, legend)
- BadgeItem rendering (locked/unlocked states, progress)
- BadgeDetailDialog (dialog content, state-dependent display)
- ProfileScreen integration

### Manual Testing Steps

1. **Open Profile Screen**
   - Navigate to profile from main screen
   - Verify badges section appears at bottom of profile

2. **Badge Grid Display**
   - Verify 4-column grid layout
   - Check "Achievements" header is visible
   - Confirm unlock counter displays (e.g., "1/3")
   - Observe locked badges have lock icon
   - Verify unlocked badges show original icon with full opacity

3. **Progress Indicators**
   - Check locked badges with >0 progress show progress bar at bottom
   - Verify progress bar color matches badge rarity
   - Confirm locked badges with 0 progress have no progress bar

4. **Rarity Legend**
   - Verify "Rarity Levels" header appears below grid
   - Check four colored dots display (grey, blue, purple, gold)
   - Confirm rarity names appear: Common, Rare, Epic, Legendary

5. **Badge Click Interaction**
   - Tap any badge → Dialog opens
   - Verify large icon displays in dialog
   - Check badge title, rarity label, and description appear
   - For unlocked badges: "Unlocked" text in rarity color
   - For locked badges with progress: Progress section with percentage
   - For locked badges without progress: "Locked" text in grey
   - Tap "Close" button → Dialog dismisses

6. **Animation Testing**
   - Observe smooth alpha animation when switching between profiles
   - Verify 300ms fade transition for locked/unlocked states

7. **Accessibility Testing**
   - Enable TalkBack / VoiceOver
   - Verify badge icons read badge title as content description
   - Check dialog elements read in logical order

---

## Screenshots

<div align="center">

<table>
  <tr>
    <td align="center" width="33%">
      Badges Screenshot 1
    </td>
    <td align="center" width="33%">
      Badges Screenshot 2
    </td>
    <td align="center" width="33%">
      Badges Screenshot 3
    </td>
  </tr>
</table>

</div>

---

## Checklist
- [x] Sufficient documentation / KDoc
- [x] All tests passing (19 new tests: 6 unit + 12 instrumentation)
- [x] No new warnings
- [x] Follows conventional commits

### Additional Notes
- Fully aligned with Material Design 3 guidelines
- Strong accessibility with comprehensive contentDescription attributes
- Reactive UI with animateFloatAsState for smooth transitions
- Reusable color gradient system for badge rarities
- Sample badges provide immediate visual feedback for development
- Future-ready for backend integration (sample data easily replaceable)

### Files Modified
**Data Layer:**
- `Badge.kt` (new) – Badge data class
- `BadgeRarity.kt` (new) – Rarity enum  
- `SampleBadges.kt` (new) – Sample badge data

**UI Layer:**
- `BadgeSection.kt` (new) – BadgesSection, BadgeItem, BadgeDetailDialog, BadgeRarityLegend composables
- `ProfileScreen.kt` (modified) – Integrated BadgesSection into ViewProfileContent

**Test Layer:**
- `BadgeTest.kt` (new) – 6 unit tests
- `BadgeSectionTest.kt` – 12 instrumentation tests

### CODE REVIEW

File: app/src/main/java/com/swent/mapin/model/badge/Badge.kt [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+ */
+data class Badge(
+    val id: String,
+    val title: String,
+    val description: String,
+    val icon: ImageVector,
+    val rarity: BadgeRarity,
+    val isUnlocked: Boolean = false,
+    val progress: Float = 0f // 0.0 to 1.0
+)
```
Comment (b10taj) [inline app/src/main/java/com/swent/mapin/model/badge/Badge.kt]: **Problem** : There is no input validation on `progress`. This can lead to corrupted state if out-of-bounds values are passed in.

**Suggestion** : Enforce **progress** in [0f, 1f], either in constructor or by using a custom getter/setter, or at point of use.

**Code suggestion** :
```
val safeProgress = progress.coerceIn(0f, 1f)
```

Or in the model : 
```
val progress: Float = progress
    get() = field.coerceIn(0f, 1f)
```

File: app/src/test/java/com/swent/mapin/model/badge/BadgeTest.kt [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+            id = "o",
+            title = "t",
+            description = "d",
+            icon = icon,
+            rarity = BadgeRarity.COMMON,
+            progress = 1f)
+
+    assertEquals(0f, zero.progress, 0.0f)
+    assertEquals(1f, one.progress, 0.0f)
+  }
```
[the remark this answers is not in the document]
Comment (naderbahri) [reply app/src/test/java/com/swent/mapin/model/badge/BadgeTest.kt]: Thanks for reminding me to add edge cases tests. I added them!

**Review by b10taj — APPROVED** [review]
Only changes are KtfmtFormat. It is approved !
